text/0000-experimental-coroutines.md

```diff
@@ -0,0 +1,471 @@
+- Feature Name: `coroutines`
+- Start Date: 2017-06-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+This is an **experimental RFC** for adding a new feature to the language,
+coroutines (also commonly referred to as generators). This RFC is intended to be
+relatively lightweight and bikeshed free as it will be followed by a separate
+RFC in the future for stabilization of this language feature. The intention here
+is to make sure everyone's on board with the *general idea* of
+coroutines/generators being added to the Rust compiler and available for use on
+the nightly channel.
+
+# Motivation
+[motivation]: #motivation
+
+One of Rust's 2017 roadmap goals is ["Rust should be well-equipped for writing
+robust, high-scale servers"][goal]. A [recent survey][survey] has shown that
+the biggest blocker to robust, high-scale servers is ergonomic usage of async
```

Comment:
Putting aside the details of how the language might expose it, is async I/O actually a limiting factor with regard to performance, or can an appropriately-designed server achieve the same performance level under a blocking model?

I've done some looking into existing literature on this question and found no meaningful conclusions either way- there seem to be many people *assuming* async I/O is always the best answer, but very little data supporting a conclusion either way.

The historical objections to threading seem to stem from

1. Some OSes historically handle large numbers of threads poorly.
2. Limited address space with fixed-size stacks. Still relevant to 32-bit systems, but most server platforms nowadays are 64-bit.
3. Concurrency is difficult to handle correctly.

The first two points seem to have become less relevant in the ~20 year history of people investigating these issues, and the third doesn't seem to make much sense on today's CPUs where we've reached the conclusion that they're not going to get much faster anymore, but they will gain higher concurrency. Good tools motivated by those observations (techniques like message passing and Rust itself) make it much easier to handle true concurrency correctly, reducing the relative benefits of event-driven async I/O which fakes it and allows programmers to ignore concurrency-related issues.

---

This question may be orthogonal to whether it's useful to have async/await support from the compiler, but I'd be hesitant to go baking in support for a feature that turns out to be unnecessary for the stated goal, given the async model is (at least in my view) is more cumbersome to write than merely blocking. If there is no meaningful benefit to this model over others, why should it be chosen as blessed to gain compiler support? The answer here may be "because it's not feasible to implement without compiler support" or "there's no technical advantage to it but this is easier than trying to change popular opinion"- the part that seems to be missing from this RFC and much discussion of async in general is justification like that.

I'm generally of the opinion that green threading (vs stackless coroutines) is a better model simply due to ergonomics (logically it's just blocking I/O which is, again in my opinion, conceptually the simplest model). There may be compelling technical reasons that stackless is worthwhile to support.